pkg/store/consul/configstore/store.go

```diff
@@ -108,24 +106,53 @@ func (cs *ConsulStore) PutConfig(ctx context.Context, config Fields, v *Version)
 	return nil
 }
 
-func (cs *ConsulStore) DeleteConfig(_ context.Context, id ID, v *Version) error {
-	kvPair := &api.KVPair{
-		Key:         id.String(),
-		ModifyIndex: v.uint64(),
+func (cs *ConsulStore) PutConfigTxn(ctx context.Context, config Fields, v *Version) error {
+	yamlConfig, err := yaml.Marshal(config.Config)
+	if err != nil {
+		return err
 	}
+	env := envelope{Config: string(yamlConfig)}
 
-	ok, _, err := cs.consulKV.DeleteCAS(kvPair, nil)
+	bs, err := json.Marshal(env)
+	err = transaction.Add(ctx, api.KVTxnOp{
+		Verb:  string(api.KVSet),
+		Key:   config.ID.String(),
+		Value: bs,
+		Index: v.uint64(),
+	})
+	if err != nil {
+		return util.Errorf("Failed to add PutConfig operation to the transaction: %v", err)
+	}
+	return nil
+}
+
+func (cs *ConsulStore) DeleteConfig(id ID, v *Version) error {
+	ok, _, err := cs.consulKV.DeleteCAS(&api.KVPair{
+		Key:         id.String(),
+		ModifyIndex: v.uint64(),
+	}, nil)
 	if !ok {
 		return util.Errorf("CAS Delete Failed! Consider retry.")
 	}
 	if err != nil {
```

Comment:
probably should check `err` first, i think `ok` will always be false when an error occurs